staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -35,34 +35,49 @@ import (
 
 // NewSerializer creates a JSON serializer that handles encoding versioned objects into the proper JSON form. If typer
 // is not nil, the object has the group, version, and kind fields set.
+// Deprecated: use NewSerializerWithOptions instead.
 func NewSerializer(meta MetaFactory, creater runtime.ObjectCreater, typer runtime.ObjectTyper, pretty bool) *Serializer {
-	return &Serializer{
-		meta:    meta,
-		creater: creater,
-		typer:   typer,
-		yaml:    false,
-		pretty:  pretty,
-	}
+	return NewSerializerWithOptions(meta, creater, typer, &SerializerOptions{false, pretty, false})
 }
 
 // NewYAMLSerializer creates a YAML serializer that handles encoding versioned objects into the proper YAML form. If typer
 // is not nil, the object has the group, version, and kind fields set. This serializer supports only the subset of YAML that
 // matches JSON, and will error if constructs are used that do not serialize to JSON.
+// Deprecated: use NewSerializerWithOptions instead.
 func NewYAMLSerializer(meta MetaFactory, creater runtime.ObjectCreater, typer runtime.ObjectTyper) *Serializer {
+	return NewSerializerWithOptions(meta, creater, typer, &SerializerOptions{true, false, false})
+}
+
+// NewSerializerWithOptions creates a JSON/YAML serializer that handles encoding versioned objects into the proper JSON/YAML
+// form. If typer is not nil, the object has the group, version, and kind fields set.
+func NewSerializerWithOptions(meta MetaFactory, creater runtime.ObjectCreater, typer runtime.ObjectTyper, serializerOptions *SerializerOptions) *Serializer {
 	return &Serializer{
-		meta:    meta,
-		creater: creater,
-		typer:   typer,
-		yaml:    true,
+		meta:              meta,
+		creater:           creater,
+		typer:             typer,
+		SerializerOptions: serializerOptions,
 	}
 }
 
+// SerializerOptions holds the options which are used to creating a JSON/YAML serializer.
+// For example:
+// (1) we can creates a JSON serializer once we set `yaml` to `false`.
+// (2) we can creates a YAML serializer once we set `yaml` to `true`. This serializer supports only the subset of YAML that
+//     matches JSON, and will error if constructs are used that do not serialize to JSON.
+// (3) we can creates a strict JSON/YAML serializer that can also return errors of type strictDecodingError, once we set
+//    `strict` to `true`. And note that this serializer is not as performant as the non-strict variant, and should not be
+//    used in fast paths.
```

Comment:
We should comment that `Pretty` is silently ignored when `YAML` is true, because YAML is always printed with white-space.

Reply:
fixed